### PR TITLE
adds support for timestamp

### PR DESCRIPTION
**Pull request by brunocalza**
With this PR, this CLI can provide a timestamp to the uploaded file, and can also specify a timestamp when listing CIDs 

The implementation is straightforward
- Adds migration
- Changes RPC definitions
- Change Upload RPC to store timestamp in GCS
- Changes Listing RPCs to filter by timestamp if provided

See also: https://github.com/tablelandnetwork/basin-storage/pull/7, https://github.com/tablelandnetwork/basin-cli/pull/16

### CODE REVIEW

File: lib/worker/src/db/publications.rs [hunk 1]
```diff
@@ -65,21 +68,16 @@ pub async fn pub_cids(
...
-                                                LIMIT $3
-                                                OFFSET $4",
-    )
-    .bind(ns)
-    .bind(rel)
-    .bind(limit)
-    .bind(offset)
-    .fetch_all(pool)
-    .await?;
+    let sql = pub_cids_build_query(&ns, &rel, &limit, &offset, &timestamp);
```
Comment (brunocalza) [inline lib/worker/src/db/publications.rs:73]: build the query for fetching cids. this query builder considers if timestamp is present or not. it is not present if it is equals `0`

File: lib/worker/src/rpc.rs [hunk 2]
```diff
@@ -148,7 +158,17 @@ impl<E: EVMClient + 'static> publications::Server for Publications<E> {
                 return Err(Error::failed("namespace not found".into()));
             }
 
-            let upload_type = UploadType::Simple(Media::new(filename.clone()));
+            let upload_type = UploadType::Multipart(Box::new(Object {
```
Comment (brunocalza) [inline lib/worker/src/rpc.rs:161]: we received the timestamp and pass as metadata to GCS

File: lib/common/migrations/20231108200529_timestamp.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE jobs ADD COLUMN timestamp BIGINT;
```
Comment (avichalp) [inline lib/common/migrations/20231108200529_timestamp.sql:1]: 👍

File: lib/worker/src/rpc.rs (hunk 2 above)
Comment (avichalp) [reply lib/worker/src/rpc.rs:161]: the cli will send current timestamp if the user doesn't provide the timestamp, right?

File: lib/worker/src/rpc.rs [hunk 4]
```diff
@@ -148,7 +158,17 @@ impl<E: EVMClient + 'static> publications::Server for Publications<E> {
...
+                name: filename.clone(),
+                content_type: Some("application/octet-stream".into()),
+                size: size as i64,
+                metadata: Some(HashMap::from([(
+                    "timestamp".into(),
+                    format!("{}", timestamp),
```
Comment (avichalp) [inline lib/worker/src/rpc.rs:167]: 👍

**Review by avichalp — APPROVED** [review]
👍

File: lib/worker/src/rpc.rs (hunk 2 above)
Comment (brunocalza) [reply lib/worker/src/rpc.rs:161]: yeah 👍

File: compose.yaml [hunk 5]
```diff
@@ -36,4 +36,4 @@ services:
     ports:
       - 26257:26257
       - 8080:8080
-    command: start-single-node --insecure --store=type=mem,size=0.25 --advertise-addr=localhost
```
Comment (sanderpick) [inline compose.yaml:39]: is a stateful node more useful here?

**Review by sanderpick — APPROVED** [review]
very nice!

File: compose.yaml (hunk 5 above)
Comment (brunocalza) [reply compose.yaml:39]: hum, i removed that because i was running out of memory every time. Maybe i should try just increasing the limit

Comment (sanderpick) [reply compose.yaml:39]: worth a try!

**Pull request merged**